File: gently/visualization/web/static/js/narrative.js
/**
 * Narrative Summary Manager
 * Manages AI-generated experiment summaries
 */

const NarrativeManager = {
    isLoading: false,
    lastNarrative: null,
    isCollapsed: false,

    init() {
        // Load collapsed state from localStorage
        const collapsed = localStorage.getItem('gently-narrative-collapsed');
        this.isCollapsed = collapsed === 'true';
        this.applyCollapseState();
    },

    toggle() {
        this.isCollapsed = !this.isCollapsed;
        localStorage.setItem('gently-narrative-collapsed', this.isCollapsed.toString());
        this.applyCollapseState();
    },

    applyCollapseState() {
        const panel = document.getElementById('narrative-panel');
        if (panel) {
            panel.classList.toggle('collapsed', this.isCollapsed);
        }
    },

    async refresh() {
        if (this.isLoading) return;

        this.isLoading = true;
        this.showLoading(true);

        try {
            const response = await fetch('/api/narrative');
            if (response.ok) {
                const narrative = await response.json();
                this.lastNarrative = narrative;
                this.renderNarrative(narrative);
            } else {
                this.renderError('Failed to generate summary');
            }
        } catch (error) {
            console.error('Failed to fetch narrative:', error);
            this.renderError('Connection error');
        } finally {
            this.isLoading = false;
            this.showLoading(false);
        }
    },

    async showSinceLastCheck() {
        if (this.isLoading) return;

        const lastCheck = ExperimentStrip.lastCheck;
        if (!lastCheck) {
            this.renderLocalSummary();
            return;
        }

        this.isLoading = true;
        this.showLoading(true);

        try {
            const response = await fetch(`/api/narrative?since=${lastCheck.toISOString()}`);
            if (response.ok) {
                const narrative = await response.json();
                this.lastNarrative = narrative;
                this.renderNarrative(narrative);
            } else {
                this.renderLocalSummary();
            }
        } catch (error) {
            this.renderLocalSummary();
        } finally {
            this.isLoading = false;
            this.showLoading(false);
            ExperimentStrip.markChecked();
        }
    },

    renderLocalSummary() {
        // Generate a simple local summary when API isn't available
        const state = EmbryosManager.state;
        const embryoCount = Object.keys(state.embryos).length;

        if (embryoCount === 0) {
            this.updateNarrativeUI({
                status: 'normal',
                headline: 'No Active Experiment',
                details: ['Start a timelapse to see AI-generated summaries here.']
            });
            return;
        }

        const activeCount = Object.values(state.embryos).filter(e => !e.isComplete).length;
        const completedCount = Object.values(state.embryos).filter(e => e.isComplete).length;

        // Check if using perception system (has stage data) or legacy detectors
        let isPerception = false;
        const stageInfo = {};  // embryo_id -> current stage
        const hatchingEmbryos = [];
        const stageOrder = ['early', 'bean', 'comma', '1.5fold', '2fold', '3fold', 'pretzel', 'hatching', 'hatched'];

        // Count stages or detections
        let totalDetections = 0;
        let detectionDetails = [];

        Object.entries(EmbryosManager.detectionReasoning).forEach(([embryoId, reasoning]) => {
            // Check for perception data
            const stages = reasoning.map(r => r.stage).filter(Boolean);
            if (stages.length > 0) {
                isPerception = true;
                stageInfo[embryoId] = stages[stages.length - 1];  // Latest stage
                if (reasoning.some(r => r.is_hatching)) {
                    hatchingEmbryos.push(embryoId);
                }
            } else {
                // Legacy detection counting
                const positives = reasoning.filter(r => r.detected);
                totalDetections += positives.length;
                positives.forEach(d => {
                    detectionDetails.push(`${embryoId}: ${EmbryosManager.formatDetectorName(d.detector_name)} at T${d.timepoint}`);
                });
            }
        });

        const details = [];
        if (activeCount > 0) details.push(`${activeCount} embryo${activeCount !== 1 ? 's' : ''} actively imaging`);
        if (completedCount > 0) details.push(`${completedCount} embryo${completedCount !== 1 ? 's' : ''} completed`);
        details.push(`${state.totalTimepoints} total timepoints acquired`);

        let status, headline;

        if (isPerception && Object.keys(stageInfo).length > 0) {
            // Show stage distribution for perception
            const stageCounts = {};
            Object.values(stageInfo).forEach(stage => {
                stageCounts[stage] = (stageCounts[stage] || 0) + 1;
            });
            const sortedStages = Object.entries(stageCounts)
                .sort((a, b) => {
                    const idxA = stageOrder.indexOf(a[0].toLowerCase());
                    const idxB = stageOrder.indexOf(b[0].toLowerCase());
                    return (idxA >= 0 ? idxA : 99) - (idxB >= 0 ? idxB : 99);
                });
            const stageSummary = sortedStages.map(([stage, count]) => `${count} ${stage}`).join(', ');
            details.push(`Stages: ${stageSummary}`);

            if (hatchingEmbryos.length > 0) {
                details.push(`Hatching detected: ${hatchingEmbryos.join(', ')}`);
                status = 'notable';
                headline = `Hatching in ${hatchingEmbryos.length} Embryo${hatchingEmbryos.length !== 1 ? 's' : ''}`;
            } else {
                // Find most advanced stage
                const maxStageIdx = Math.max(...Object.values(stageInfo).map(s => {
                    const idx = stageOrder.indexOf(s.toLowerCase());
                    return idx >= 0 ? idx : 0;
                }));
                const maxStage = stageOrder[maxStageIdx].replace('fold', '-fold');
                status = 'normal';
                headline = `Most Advanced: ${maxStage.charAt(0).toUpperCase() + maxStage.slice(1)}`;
            }
        } else if (detectionDetails.length > 0) {
            details.push(`${totalDetections} positive detection${totalDetections !== 1 ? 's' : ''}: ${detectionDetails.slice(0, 3).join(', ')}${detectionDetails.length > 3 ? '...' : ''}`);
            status = 'notable';
            headline = `${totalDetections} Detection${totalDetections !== 1 ? 's' : ''} Found`;
        } else if (completedCount > 0) {
            status = 'normal';
            headline = `${completedCount}/${embryoCount} Embryos Complete`;
        } else {
            status = 'normal';
            headline = 'Experiment In Progress';
        }

        this.updateNarrativeUI({ status, headline, details });
    },

    renderNarrative(narrative) {
        this.updateNarrativeUI({
            status: narrative.status || 'normal',
            headline: narrative.headline || 'Experiment Summary',
            summary: narrative.summary,
            details: narrative.details || []
        });
    },

    renderError(message) {
        this.updateNarrativeUI({
            status: 'attention',
            headline: 'Summary Unavailable',
            details: [message, 'Showing local summary instead.']
        });
        // Fall back to local summary
        setTimeout(() => this.renderLocalSummary(), 2000);
    },

    updateNarrativeUI({ status, headline, summary, details }) {
        // Update badge
        const badge = document.getElementById('narrative-badge');
        if (badge) {
            badge.className = `narrative-status-badge ${status}`;
            badge.innerHTML = status === 'normal' ? '&#x2714;' :
                             status === 'notable' ? '&#x1F514;' : '&#x26A0;';
        }

        // Update headline
        const headlineEl = document.getElementById('narrative-headline');
        if (headlineEl) headlineEl.textContent = headline;

        // Update meta
        const metaEl = document.getElementById('narrative-meta');
        if (metaEl) metaEl.textContent = `Updated ${new Date().toLocaleTimeString()}`;

        // Update body
        const bodyEl = document.getElementById('narrative-body');
        if (bodyEl) {
            let html = '';

            if (summary) {
                html += `
                    <div class="narrative-section">
                        <div class="narrative-section-title">Summary</div>
                        <p class="narrative-text">${this.escapeHtml(summary)}</p>
                    </div>
                `;
            }

            if (details && details.length > 0) {
                html += `
                    <div class="narrative-section">
                        <div class="narrative-section-title">Details</div>
                        <ul class="narrative-bullets">
                            ${details.map(d => `<li>${this.escapeHtml(d)}</li>`).join('')}
                        </ul>
                    </div>
                `;
            }

            bodyEl.innerHTML = html || '<p class="narrative-text">No summary available.</p>';
        }
    },

    showLoading(show) {
        const loading = document.getElementById('narrative-loading');
        const body = document.getElementById('narrative-body');
        if (loading) loading.style.display = show ? 'flex' : 'none';
        if (body) body.style.display = show ? 'none' : 'block';
    },

    escapeHtml(text) {
        const div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    }
};

// Initialize on DOM ready
document.addEventListener('DOMContentLoaded', () => {
    NarrativeManager.init();
    // Generate initial narrative summary
    setTimeout(() => NarrativeManager.renderLocalSummary(), 500);
});

File: gently/ui/web/narrative.py
"""
Narrative Generation for the Visualization Server
===================================================

Generates experiment narrative summaries from timelapse tracker state.
"""

from datetime import datetime
from typing import Any, Dict, Optional

from .timelapse_tracker import TimelapseStateTracker


def generate_narrative_summary(
    tracker: TimelapseStateTracker,
    since: Optional[str] = None,
) -> Dict[str, Any]:
    """Generate a narrative summary of the experiment state.

    This generates a local summary based on the timelapse tracker state.
    Can be extended to use Claude Haiku for AI-powered narratives.

    Args:
        tracker: The timelapse state tracker
        since: Optional ISO timestamp to generate differential summary

    Returns:
        Dict with status, headline, summary, and details
    """
    # If no active experiment
    if tracker.status == "IDLE" and not tracker.embryos:
        return {
            "status": "normal",
            "headline": "No Active Experiment",
            "summary": None,
            "details": ["Start a timelapse to see experiment summaries here."],
            "generated_at": datetime.now().isoformat()
        }

    # Count embryos
    embryo_count = len(tracker.embryos)
    active_embryos = [e for e in tracker.embryos.values() if not e.get("is_complete")]
    completed_embryos = [e for e in tracker.embryos.values() if e.get("is_complete")]

    # Check if using perception system (has stage data) or legacy detectors
    is_perception = False
    stage_info = {}  # embryo_id -> current stage
    hatching_embryos = []
    stage_order = ['early', 'bean', 'comma', '1.5fold', '2fold', '3fold', 'pretzel', 'hatching', 'hatched']

    for embryo_id, reasoning_list in tracker.detection_reasoning.items():
        # Check for stage data (perception system)
        stages = [r.get("stage") for r in reasoning_list if r.get("stage")]
        if stages:
            is_perception = True
            stage_info[embryo_id] = stages[-1]  # Latest stage
            # Check for hatching
            if any(r.get("is_hatching") for r in reasoning_list):
                hatching_embryos.append(embryo_id)

    # Count detections (legacy) or stage progression (perception)
    total_detections = 0
    detection_details = []

    if is_perception:
        # Group embryos by stage for perception
        for embryo_id, stage in stage_info.items():
            detection_details.append(f"{embryo_id}: {stage.replace('fold', '-fold').title()}")
    else:
        # Legacy detection counting
        for embryo_id, reasoning_list in tracker.detection_reasoning.items():
            positives = [r for r in reasoning_list if r.get("detected")]
            total_detections += len(positives)
            for d in positives:
                detector = d.get("detector_name", "unknown")
                tp = d.get("timepoint", "?")
                detection_details.append(f"{embryo_id}: {detector.title()} at T{tp}")

    # Build details list
    details = []

    if len(active_embryos) > 0:
        details.append(f"{len(active_embryos)} embryo{'s' if len(active_embryos) != 1 else ''} actively imaging")

    if len(completed_embryos) > 0:
        details.append(f"{len(completed_embryos)} embryo{'s' if len(completed_embryos) != 1 else ''} completed")

    details.append(f"{tracker.total_timepoints} total timepoints acquired")

    if tracker.base_interval:
        interval_str = f"{tracker.base_interval // 60} min" if tracker.base_interval >= 60 else f"{tracker.base_interval}s"
        details.append(f"Imaging interval: {interval_str}")

    if is_perception and stage_info:
        # Show stage distribution for perception
        stage_counts = {}
        for stage in stage_info.values():
            stage_counts[stage] = stage_counts.get(stage, 0) + 1
        # Sort by stage order
        sorted_stages = sorted(stage_counts.items(),
                               key=lambda x: stage_order.index(x[0].lower()) if x[0].lower() in stage_order else 99)
        stage_summary = ", ".join(f"{count} {stage}" for stage, count in sorted_stages)
        details.append(f"Stages: {stage_summary}")
        if hatching_embryos:
            details.append(f"Hatching detected: {', '.join(hatching_embryos)}")
    elif detection_details:
        if len(detection_details) <= 3:
            details.append(f"{total_detections} detection{'s' if total_detections != 1 else ''}: {', '.join(detection_details)}")
        else:
            details.append(f"{total_detections} detection{'s' if total_detections != 1 else ''}: {', '.join(detection_details[:3])}...")

    # Calculate duration
    if tracker.started_at:
        started = datetime.fromisoformat(tracker.started_at) if isinstance(tracker.started_at, str) else tracker.started_at
        duration_sec = (datetime.now() - started).total_seconds()
        hours = int(duration_sec // 3600)
        minutes = int((duration_sec % 3600) // 60)
        duration_str = f"{hours}h {minutes}m" if hours > 0 else f"{minutes}m"
        details.append(f"Running for {duration_str}")

    # Determine status and headline
    if is_perception:
        if hatching_embryos:
            status = "notable"
            headline = f"Hatching Detected in {len(hatching_embryos)} Embryo{'s' if len(hatching_embryos) != 1 else ''}"
        elif stage_info:
            # Find most advanced stage
            max_stage_idx = max(stage_order.index(s.lower()) if s.lower() in stage_order else 0
                                for s in stage_info.values())
            max_stage = stage_order[max_stage_idx].replace('fold', '-fold').title()
            status = "normal"
            headline = f"Most Advanced: {max_stage}"
        else:
            status = "normal"
            headline = "Experiment In Progress"
    elif total_detections > 0:
        status = "notable"
        headline = f"{total_detections} Detection{'s' if total_detections != 1 else ''} Found"
    elif len(completed_embryos) > 0:
        status = "normal"
        headline = f"{len(completed_embryos)}/{embryo_count} Embryos Complete"
    else:
        status = "normal"
        headline = "Experiment In Progress"

    # Build summary text
    summary = None
    if is_perception:
        if hatching_embryos:
            summary = f"Hatching has been detected in {', '.join(hatching_embryos)}. Monitoring continues for all embryos."
        elif stage_info:
            # Summarize stage distribution
            unique_stages = set(stage_info.values())
            if len(unique_stages) == 1:
                summary = f"All {len(stage_info)} embryos are at {list(unique_stages)[0].replace('fold', '-fold').title()} stage."
            else:
                summary = f"Embryos are progressing through developmental stages. {len(stage_info)} embryos tracked."
    elif total_detections > 0:
        latest = detection_details[-1] if detection_details else None
        summary = f"Positive detections have been identified. {latest}. All imaging continues normally."
    elif len(completed_embryos) > 0:
        summary = f"{len(completed_embryos)} embryo{'s have' if len(completed_embryos) != 1 else ' has'} reached their stop condition. {len(active_embryos)} still being imaged."

    return {
        "status": status,
        "headline": headline,
        "summary": summary,
        "details": details,
        "generated_at": datetime.now().isoformat()
    }
